src/deluge/gui/ui/keyboard/layout/chord_keyboard.cpp

```diff
@@ -0,0 +1,142 @@
+	currentNotesState = NotesState{}; // Erase active notes
+	KeyboardStateChord& state = getState().chord;
+
+	// We run through the presses in reverse order to display the most recent pressed chord on top
+	for (int32_t idxPress = kMaxNumKeyboardPadPresses - 1; idxPress >= 0; --idxPress) {
+
+		PressedPad pressed = presses[idxPress];
+		if (pressed.active && pressed.x < kDisplayWidth) {
+
+			int32_t chordNo = pressed.y + state.chordList.chordRowOffset;
+
+			Voicing voicing = state.chordList.getChordVoicing(chordNo);
+			drawChordName(noteFromCoords(pressed.x), state.chordList.chords[chordNo].name, voicing.supplementalName);
+
+			for (int i = 0; i < kMaxChordKeyboardSize; i++) {
+				int32_t offset = voicing.offsets[i];
+				if (offset == NONE) {
+					continue;
+				}
+				enableNote(noteFromCoords(pressed.x) + offset, velocity);
+			}
+		}
+	}
+	ColumnControlsKeyboard::evaluatePads(presses);
+}
+
+void KeyboardLayoutChord::handleVerticalEncoder(int32_t offset) {
+	KeyboardStateChord& state = getState().chord;
+
+
+	state.chordList.adjustChordRowOffset(offset);
+	precalculate();
+}
+
+void KeyboardLayoutChord::handleHorizontalEncoder(int32_t offset, bool shiftEnabled,
+                                                  PressedPad presses[kMaxNumKeyboardPadPresses], bool encoderPressed) {
+	KeyboardStateChord& state = getState().chord;
+
+	if (encoderPressed) {
+		for (int32_t idxPress = kMaxNumKeyboardPadPresses - 1; idxPress >= 0; --idxPress) {
+
+			PressedPad pressed = presses[idxPress];
+			if (pressed.active && pressed.x < kDisplayWidth) {
+
+				int32_t chordNo = pressed.y + state.chordList.chordRowOffset;
+
+				state.chordList.adjustVoicingOffset(chordNo, offset);
+			}
+		}
+	}
+	else {
+		state.noteOffset += offset;
+	}
+	precalculate();
+}
+
+void KeyboardLayoutChord::precalculate() {
+	KeyboardStateChord& state = getState().chord;
+
+	// Pre-Buffer colours for next renderings
+	for (int32_t i = 0; i < (kOctaveSize + kDisplayWidth); ++i) {
+		noteColours[i] = getNoteColour(((state.noteOffset + i) % state.rowInterval) * state.rowColorMultiplier);
+	}
+}
+
+void KeyboardLayoutChord::renderPads(RGB image[][kDisplayWidth + kSideBarWidth]) {
+	// Iterate over grid image
+	for (int32_t y = 0; y < kDisplayHeight; ++y) {
+		for (int32_t x = 0; x < kDisplayWidth; x++) {
+			image[y][x] = noteColours[x];
+		}
+	}
+}
+
+void KeyboardLayoutChord::drawChordName(int16_t noteCode, const char* chordName, const char* voicingName) {
+	char noteName[3];
+	uint8_t drawDot;
+	char* thisChar = noteName;
+
+	int32_t noteCodeWithinOctave = (uint16_t)(noteCode + 120) % (uint8_t)12;
+	*thisChar = noteCodeToNoteLetter[noteCodeWithinOctave];
+	thisChar++;
+
+	if (noteCodeIsSharp[noteCodeWithinOctave]) {
+		*thisChar = display->haveOLED() ? '#' : '.';
+		thisChar++;
+		drawDot = 255;
+	}
+	else {
+		drawDot = 0;
+	}
+	*thisChar = '\0';
+
+	// Create a buffer to hold the full chord name and format it
+	char fullChordName[300];
+	if (voicingName[0] == '\0') {
```

Comment:
just 0 is usually better, it avoids potential mistakes like forgetting an escape and using '0'